packager/packager.cc

```diff
@@ -1739,30 +1751,85 @@ class ComputePackageSCCs {
             // top of the stack are in the same SCC. Pop the stack until we reach the root of the SCC, and assign them
             // the same SCC ID.
             core::packages::MangledName poppedPkgName;
-            const auto sccId = this->nextSCCId++;
+            auto &node = this->condensation.pushNode(edgeType);
+            auto sccId = node.id;
+
+            // Set the SCC ids for all of the members of the SCC
             do {
                 poppedPkgName = this->stack.back();
+                node.members.push_back(poppedPkgName);
                 this->stack.pop_back();
                 this->nodeMap[poppedPkgName].onStack = false;
-                auto &poppedPkgInfo = PackageInfoImpl::from(*(gs.packageDB().getPackageInfoNonConst(poppedPkgName)));
-                poppedPkgInfo.sccID_ = sccId;
+
+                auto &poppedPkgInfo = PackageInfoImpl::from(*(packageDB.getPackageInfoNonConst(poppedPkgName)));
+                switch (edgeType) {
+                    case core::packages::ImportType::Normal: {
+                        poppedPkgInfo.sccID_ = sccId;
+                        break;
+                    }
+
+                    case core::packages::ImportType::Test: {
+                        poppedPkgInfo.testSccID_ = sccId;
+
+                        // Tests have an implicit dependency on their package's application code. Those scc ids must
+                        // exist at this point, as we've already traversed all packages once.
+                        auto appSccId = poppedPkgInfo.sccID_;
+                        ENFORCE(appSccId.has_value());
+                        node.imports.insert(*appSccId);
+                        break;
+                    }
+                }
             } while (poppedPkgName != pkgName);
+
+            // Iterate the imports of each member, building the edges of the condensation. This step is performed after
+            // we've visited all members of the SCC once, to ensure that their ids have all been populated.
+            for (auto name : node.members) {
```

Comment:
This process is a little bit expensive right now, but I think that it will get cheaper once we've changed the storage for packages to look more like the rest of the symbol table. (Indexing into a vector is cheaper than looking up in a hash map.)